File: ts/src/symbols.ts
/* Copyright (c) 2026 Richard Rodger and contributors, MIT License */

// Symbol and macro tables used by the C parser to resolve the classic
// identifier/typedef-name disambiguation problem.
//
// Both tables live on Context.meta so lex matchers and rule actions can read
// and mutate the same instance. The identifier-classification matcher
// consults SymbolTable.isTypedef(name) before deciding whether to emit
// ID or TYPEDEF_NAME, and consults MacroTable.has(name) to optionally
// emit MACRO_NAME (which the grammar still treats as ID for syntax but
// tags downstream so call_expr.isMacro can be set).
//
// Scopes form a stack; the parser pushes/pops on:
//   - file               (root, never popped)
//   - function-prototype (function declarator with parameter list)
//   - function-body      (the compound-statement of a function definition)
//   - block              (compound-statement)
//   - struct-or-union    (member declaration list)
//   - enum               (enumerator list)
//   - for-init           (the init clause of a for loop, C99+)
//
// Inner scopes shadow outer ones; lookups walk the stack outward.

export type ScopeKind =
  | 'file'
  | 'fn-proto'
  | 'fn-body'
  | 'block'
  | 'struct-union'
  | 'enum'
  | 'for-init'

export interface Scope {
  kind: ScopeKind
  // Names bound in this scope to their kind. We track typedef-ness primarily;
  // ordinary identifiers (variables, functions, enumerators) are recorded so
  // that an inner non-typedef declaration of the same name correctly hides an
  // outer typedef.
  bind: Map<string, 'typedef' | 'ordinary'>
  // Tags (struct/union/enum names) live in a separate namespace in C.
  tags: Map<string, 'struct' | 'union' | 'enum'>
}

export class SymbolTable {
  private stack: Scope[] = [{
    kind: 'file',
    bind: new Map(),
    tags: new Map(),
  }]

  enter(kind: ScopeKind): void {
    this.stack.push({ kind, bind: new Map(), tags: new Map() })
  }

  exit(): void {
    if (this.stack.length <= 1) return // never pop the file scope
    this.stack.pop()
  }

  depth(): number {
    return this.stack.length
  }

  current(): Scope {
    return this.stack[this.stack.length - 1]
  }

  // True if `name` resolves to a typedef-name in any visible scope, with
  // the innermost binding winning.
  isTypedef(name: string): boolean {
    for (let i = this.stack.length - 1; i >= 0; i--) {
      const b = this.stack[i].bind.get(name)
      if (b !== undefined) return b === 'typedef'
    }
    return false
  }

  // True if `name` is bound at all (typedef or ordinary).
  isBound(name: string): boolean {
    for (let i = this.stack.length - 1; i >= 0; i--) {
      if (this.stack[i].bind.has(name)) return true
    }
    return false
  }

  bindTypedef(name: string): void {
    this.current().bind.set(name, 'typedef')
  }

  bindOrdinary(name: string): void {
    this.current().bind.set(name, 'ordinary')
  }

  bindTag(name: string, kind: 'struct' | 'union' | 'enum'): void {
    this.current().tags.set(name, kind)
  }

  hasTag(name: string): boolean {
    for (let i = this.stack.length - 1; i >= 0; i--) {
      if (this.stack[i].tags.has(name)) return true
    }
    return false
  }
}

// Macro definitions captured from #define directives. We do not expand;
// we only need to know which names are macros so call sites can be tagged.
export interface MacroDef {
  name: string
  isFunctionLike: boolean
  params?: string[]      // present iff isFunctionLike
  variadic?: boolean     // ... in parameter list
  // Body is preserved as a raw token sequence on the AST node; we only need
  // names here.
}

export class MacroTable {
  private defs = new Map<string, MacroDef>()

  define(def: MacroDef): void {
    this.defs.set(def.name, def)
  }

  undefine(name: string): void {
    this.defs.delete(name)
  }

  has(name: string): boolean {
    return this.defs.has(name)
  }

  get(name: string): MacroDef | undefined {
    return this.defs.get(name)
  }
}

// Lex-mode flags that ride on Context.meta. Lex matchers read these to know
// whether to enter directive-line mode (where newline becomes a token and
// header-name forms are valid) or normal mode.
export interface LexMode {
  // True between the start-of-logical-line `#` and the terminating newline.
  inDirective: boolean
  // True when the next token inside a directive should be parsed as a
  // header-name (set by the #include matcher upon seeing `include`).
  expectHeaderName: boolean
  // The directive name currently being parsed, or null. Used by matchers
  // that need to know which directive they are inside.
  directiveName: string | null
}

// All C-parser global state lives here, accessible from any matcher or rule
// via ctx.meta.cmeta.
export interface CMeta {
  symbols: SymbolTable
  macros: MacroTable
  mode: LexMode
  // Buffer of trivia tokens (comments, line continuations) emitted since
  // the last non-trivia token. The sub-lex hook drains this buffer onto
  // `tkn.use.leading` of each non-trivia token, so trivia stays attached
  // to the token it precedes even though the parser sees it as IGNORE'd.
  pendingTrivia: any[]
}

export function makeCMeta(): CMeta {
  return {
    symbols: new SymbolTable(),
    macros: new MacroTable(),
    mode: { inDirective: false, expectHeaderName: false, directiveName: null },
    pendingTrivia: [],
  }
}

File: go/symbols.go
/* Copyright (c) 2026 Richard Rodger and contributors, MIT License */

package tabnasc

// Symbol and macro tables used by the C parser to resolve the classic
// identifier/typedef-name disambiguation problem. Port of ../ts/src/symbols.ts.
//
// Both tables live on the per-parse CMeta (stashed on the engine Context) so
// lex matchers and rule actions read and mutate the same instance. The
// identifier-classification matcher consults SymbolTable.IsTypedef before
// deciding ID vs TYPEDEF_NAME, and MacroTable.Has to optionally emit
// MACRO_NAME.
//
// Scopes form a stack; the parser pushes/pops on file, function-prototype,
// function-body, block, struct-or-union, enum and for-init boundaries. Inner
// scopes shadow outer ones; lookups walk the stack outward.

// ScopeKind enumerates the C scope kinds.
type ScopeKind string

const (
	ScopeFile        ScopeKind = "file"
	ScopeFnProto     ScopeKind = "fn-proto"
	ScopeFnBody      ScopeKind = "fn-body"
	ScopeBlock       ScopeKind = "block"
	ScopeStructUnion ScopeKind = "struct-union"
	ScopeEnum        ScopeKind = "enum"
	ScopeForInit     ScopeKind = "for-init"
)

// BindKind records whether a name is bound as a typedef or an ordinary name.
type BindKind string

const (
	BindTypedef  BindKind = "typedef"
	BindOrdinary BindKind = "ordinary"
)

// TagKind records the kind of a struct/union/enum tag.
type TagKind string

const (
	TagStruct TagKind = "struct"
	TagUnion  TagKind = "union"
	TagEnum   TagKind = "enum"
)

// Scope is one lexical scope frame.
type Scope struct {
	Kind ScopeKind
	// Bind maps names bound in this scope to their kind. Ordinary identifiers
	// are recorded so an inner non-typedef declaration hides an outer typedef.
	Bind map[string]BindKind
	// Tags (struct/union/enum names) live in a separate namespace in C.
	Tags map[string]TagKind
}

func newScope(kind ScopeKind) *Scope {
	return &Scope{
		Kind: kind,
		Bind: map[string]BindKind{},
		Tags: map[string]TagKind{},
	}
}

// SymbolTable is the scope stack.
type SymbolTable struct {
	stack []*Scope
}

// NewSymbolTable creates a table with the (never-popped) file scope.
func NewSymbolTable() *SymbolTable {
	return &SymbolTable{stack: []*Scope{newScope(ScopeFile)}}
}

// Enter pushes a new scope.
func (st *SymbolTable) Enter(kind ScopeKind) {
	st.stack = append(st.stack, newScope(kind))
}

// Exit pops the current scope (never the file scope).
func (st *SymbolTable) Exit() {
	if len(st.stack) <= 1 {
		return
	}
	st.stack = st.stack[:len(st.stack)-1]
}

// Depth returns the number of scopes on the stack.
func (st *SymbolTable) Depth() int {
	return len(st.stack)
}

// Current returns the innermost scope.
func (st *SymbolTable) Current() *Scope {
	return st.stack[len(st.stack)-1]
}

// IsTypedef reports whether name resolves to a typedef-name in any visible
// scope, innermost binding winning.
func (st *SymbolTable) IsTypedef(name string) bool {
	for i := len(st.stack) - 1; i >= 0; i-- {
		if b, ok := st.stack[i].Bind[name]; ok {
			return b == BindTypedef
		}
	}
	return false
}

// IsBound reports whether name is bound at all (typedef or ordinary).
func (st *SymbolTable) IsBound(name string) bool {
	for i := len(st.stack) - 1; i >= 0; i-- {
		if _, ok := st.stack[i].Bind[name]; ok {
			return true
		}
	}
	return false
}

// BindTypedef binds name as a typedef in the current scope.
func (st *SymbolTable) BindTypedef(name string) {
	st.Current().Bind[name] = BindTypedef
}

// BindOrdinary binds name as an ordinary identifier in the current scope.
func (st *SymbolTable) BindOrdinary(name string) {
	st.Current().Bind[name] = BindOrdinary
}

// BindTag binds a struct/union/enum tag in the current scope.
func (st *SymbolTable) BindTag(name string, kind TagKind) {
	st.Current().Tags[name] = kind
}

// HasTag reports whether a tag with name is visible.
func (st *SymbolTable) HasTag(name string) bool {
	for i := len(st.stack) - 1; i >= 0; i-- {
		if _, ok := st.stack[i].Tags[name]; ok {
			return true
		}
	}
	return false
}

// MacroDef is a captured #define. We do not expand; we only track names so
// call sites can be tagged.
type MacroDef struct {
	Name           string
	IsFunctionLike bool
	Params         []string // present iff IsFunctionLike
	Variadic       bool     // ... in parameter list
}

// MacroTable records #define'd names.
type MacroTable struct {
	defs map[string]*MacroDef
}

// NewMacroTable creates an empty macro table.
func NewMacroTable() *MacroTable {
	return &MacroTable{defs: map[string]*MacroDef{}}
}

// Define records a macro.
func (mt *MacroTable) Define(def *MacroDef) {
	mt.defs[def.Name] = def
}

// Undefine removes a macro.
func (mt *MacroTable) Undefine(name string) {
	delete(mt.defs, name)
}

// Has reports whether name is a defined macro.
func (mt *MacroTable) Has(name string) bool {
	_, ok := mt.defs[name]
	return ok
}

// Get returns the macro def for name, or nil.
func (mt *MacroTable) Get(name string) *MacroDef {
	return mt.defs[name]
}

// LexMode are lex-mode flags that ride on CMeta. Lex matchers read these to
// know whether to enter directive-line mode (where newline becomes a token
// and header-name forms are valid) or normal mode.
type LexMode struct {
	// InDirective is true between the start-of-logical-line `#` and the
	// terminating newline.
	InDirective bool
	// ExpectHeaderName is true when the next token inside a directive should
	// be parsed as a header-name (set by the #include matcher).
	ExpectHeaderName bool
	// DirectiveName is the directive currently being parsed, or "".
	DirectiveName string
}

// CMeta holds all C-parser per-parse state, accessible from any matcher or
// rule via the engine Context.
type CMeta struct {
	Symbols *SymbolTable
	Macros  *MacroTable
	Mode    *LexMode
	// PendingTrivia buffers trivia tokens (comments, line continuations)
	// emitted since the last non-trivia token. The sub-lex hook drains this
	// onto each non-trivia token's leading-trivia so trivia stays attached to
	// the token it precedes even though the parser IGNOREs it.
	PendingTrivia []any
}

// MakeCMeta builds a fresh CMeta.
func MakeCMeta() *CMeta {
	return &CMeta{
		Symbols:       NewSymbolTable(),
		Macros:        NewMacroTable(),
		Mode:          &LexMode{},
		PendingTrivia: nil,
	}
}
